.github/workflows/ci.yml

```diff
@@ -13,89 +13,164 @@ env:
-  WASM_EXAMPLES_LIST: "browser-echo"
+  RS_EXAMPLES_LIST: "browser-chat browser-chat/cli browser-echo custom-router dumbpipe-web framed-messages frosty iroh-automerge iroh-automerge-repo iroh-gateway extism/host extism/plugin extism/iroh-extism-host-functions"
+  WASM_EXAMPLES_LIST: "browser-chat browser-echo"
   IROH_FORCE_STAGING_RELAYS: "1"
 
 jobs:
-  build_and_test_nix:
-    timeout-minutes: 30
-    name: Build and test (Nix)
-    runs-on: ${{ matrix.runner }}
+  # Prepare dependencies and cache them for all other jobs
+  prepare:
+    name: Prepare dependencies
+    runs-on: ubuntu-latest
+    timeout-minutes: 20
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          submodules: recursive
+
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          components: clippy,rustfmt
+
+      - name: Add wasm target
+        run: rustup target add wasm32-unknown-unknown
+
+      - name: Setup Rust cache
+        uses: Swatinem/rust-cache@v2
+        with:
+          shared-key: "iroh-examples-shared"
+          cache-all-crates: true
+
+      - name: Fetch dependencies
+        run: |
+          for example in $RS_EXAMPLES_LIST; do
+            if [ -f "$example/Cargo.toml" ]; then
+              echo "Fetching dependencies for $example"
+              cargo fetch --manifest-path "$example/Cargo.toml" --locked 2>/dev/null || cargo fetch --manifest-path "$example/Cargo.toml"
+            fi
+          done
+
+  # Check formatting for all examples in a single job (fast)
+  fmt:
+    name: Format check
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          submodules: recursive
+
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          components: rustfmt
+
+      - name: Check formatting
+        run: |
+          for example in $RS_EXAMPLES_LIST; do
+            if [ -f "$example/Cargo.toml" ]; then
+              echo "Checking format: $example"
+              cargo fmt --manifest-path "$example/Cargo.toml" --all -- --check
+            fi
+          done
+
+  # Build WASM examples
+  wasm:
+    name: WASM build
+    runs-on: ubuntu-latest
+    timeout-minutes: 20
+    needs: prepare
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          submodules: recursive
+
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@stable
+
+      - name: Add wasm target
+        run: rustup target add wasm32-unknown-unknown
+
+      - name: Setup Rust cache
+        uses: Swatinem/rust-cache@v2
+        with:
+          shared-key: "iroh-examples-shared"
+
+      - name: Build WASM examples
+        run: |
+          # browser-echo is a single package
```

Comment:
can we use the WASM_EXAMPLES_LIST from above here instead of manually invoking the build for each wasm crate?